openedx/core/djangoapps/password_policy/compliance.py

```diff
@@ -0,0 +1,136 @@
+    user.
+
+    Important: This method should only be called AFTER the user has been authenticated.
+    """
+    is_compliant = _check_user_compliance(user, password)
+    if is_compliant:
+        return
+
+    deadline = _get_compliance_deadline_for_user(user)
+    if deadline is None:
+        return
+
+    now = datetime.now(pytz.UTC)
+    if now >= deadline:
+        raise NonCompliantPasswordException(
+            _(
+                '{platform_name} now requires more complex passwords. Your current password does not meet the new '
+                'requirements. Change your password now to continue using the site. Thank you for helping us keep '
+                'your data safe.'
+            ).format(platform_name=settings.PLATFORM_NAME)
+        )
+    else:
+        raise NonCompliantPasswordWarning(
+            _(
+                '{platform_name} now requires more complex passwords. Your current password does not meet the new '
+                'requirements. You must change your password by {deadline} to be able to continue using the site. '
+                'Thank you for helping us keep your data safe.'
+            ).format(
+                platform_name=settings.PLATFORM_NAME,
+                deadline=strftime_localized(deadline, DEFAULT_SHORT_DATE_FORMAT)
+            )
+        )
+
+
+def _rollout_config():
+    """
+    Return a dictionary with configuration settings for managing the rollout of password policy compliance
+    enforcement.
+    """
+    return getattr(settings, 'PASSWORD_POLICY_COMPLIANCE_ROLLOUT_CONFIG', {})
+
+
+def _check_user_compliance(user, password):
+    """
+    Returns a boolean indicating whether or not the user is compliant with password policy rules.
+    """
+    try:
+        validate_password(password, user=user, password_reset=False)
+        return True
+    except Exception:  # pylint: disable=broad-except
+        # If anything goes wrong, we should assume the password is not compliant but we don't necessarily
+        # need to prevent login.
+        return False
+
+
+def _get_compliance_deadline_for_user(user):
+    """
+    Returns the date that the user will be required to comply with password policy rules, or None if no such date
+    applies to this user. If a deadline is not set, it will fall back to a more general deadline that is set.
+    """
+    config = _rollout_config()
+
+    # Implied hierarchy of general->staff in terms of scope, so we'll use each as a fallback to the other for any
+    # blank fields.
+    general_deadline = config.get('GENERAL_USER_COMPLIANCE_DEADLINE', None)
+    privilege_deadline = config.get('ELEVATED_PRIVILEGE_USER_COMPLIANCE_DEADLINE', general_deadline)
+    staff_deadline = config.get('STAFF_USER_COMPLIANCE_DEADLINE', privilege_deadline)
+
+    # Only keep and parse deadlines that matter to this user
+    staff_deadline = staff_deadline and user.is_staff and parse_date(staff_deadline)
```

Comment:
Is there a way we can catch `parse_date` errors from malformed dates earlier in the process or `except` the expected exceptions so we don't blow up all our users before we find out